File: client/src/main/java/net/hoz/netapi/client/service/NetPlayerProvider.java
package net.hoz.netapi.client.service;

import com.github.benmanes.caffeine.cache.Cache;
import com.github.benmanes.caffeine.cache.Caffeine;
import com.google.inject.Inject;
import com.google.protobuf.Empty;
import com.iamceph.resulter.core.DataResultable;
import com.iamceph.resulter.core.Resultable;
import lombok.Getter;
import lombok.experimental.Accessors;
import lombok.extern.slf4j.Slf4j;
import net.hoz.api.Controlled;
import net.hoz.api.data.NetPlayer;
import net.hoz.api.data.NetPlayerHistory;
import net.hoz.api.data.PlayerSettings;
import net.hoz.api.data.WUUID;
import net.hoz.api.service.NetPlayerRequest;
import net.hoz.api.service.NetPlayerServiceClient;
import net.hoz.api.util.Packeto;
import net.hoz.api.util.ReactorHelper;
import net.hoz.netapi.client.util.NetUtils;
import org.screamingsandals.lib.utils.Controllable;
import reactor.core.Disposable;
import reactor.core.publisher.Flux;
import reactor.core.publisher.Mono;
import reactor.core.publisher.Sinks;

import java.time.Duration;
import java.util.Locale;
import java.util.UUID;

/**
 * Service for managing generally players.
 */
@Accessors(fluent = true)
@Slf4j
public class NetPlayerProvider implements Controlled {
    /**
     * Cache of the player data.
     */
    private final Cache<UUID, NetPlayer> playerCache = Caffeine.newBuilder()
            .expireAfterAccess(Duration.ofMinutes(30))
            .build();
    /**
     * Cache of the player history.
     */
    private final Cache<UUID, NetPlayerHistory> historyCache = Caffeine.newBuilder()
            .expireAfterAccess(Duration.ofMinutes(30))
            .build();

    private final Sinks.Many<NetPlayer> playerUpdater;

    /**
     * RSocket service.
     */
    private final NetPlayerServiceClient netPlayerService;

    @Getter
    private final Settings settings;

    private Disposable updateListener;

    @Inject
    public NetPlayerProvider(NetPlayerServiceClient netPlayerService) {
        this.netPlayerService = netPlayerService;
        this.settings = new Settings(this);
        this.playerUpdater = Sinks.many().multicast().directBestEffort();
    }

    @Override
    public void initialize() {
        subscribeToUpdates();
    }

    @Override
    public void dispose() {
        if (updateListener != null) {
            updateListener.dispose();
        }

        playerCache.invalidateAll();
        historyCache.invalidateAll();
    }

    /**
     * Tries to get the {@link NetPlayer} from the cache.
     *
     * @param uuid ID of the player
     * @return {@link DataResultable} result of the operation.
     */
    public DataResultable<NetPlayer> getPlayer(UUID uuid) {
        return DataResultable.failIfNull(playerCache.getIfPresent(uuid));
    }

    /**
     * Tries to get the {@link NetPlayerHistory} from the cache.
     *
     * @param uuid ID of the player
     * @return {@link DataResultable} result of the operation.
     */
    public DataResultable<NetPlayerHistory> getHistory(UUID uuid) {
        return DataResultable.failIfNull(historyCache.getIfPresent(uuid));
    }

    /**
     * Tries to get the {@link NetPlayer} data container from the BAGR instance.
     *
     * @param uuid
     * @param address
     * @return
     */
    public Mono<DataResultable<NetPlayer>> loadPlayer(UUID uuid, String address) {
        return netPlayerService.dataFor(NetPlayerRequest.newBuilder()
                        .setUUID(uuid.toString())
                        .setAddress(address)
                        .build())
                .filter(ReactorHelper.filterResult(log))
                .map(data -> Packeto.unpack(data, NetPlayer.class))
                .doOnNext(result ->
                        result.ifOk(data -> playerCache.put(uuid, data)))
                .onErrorResume(ex -> ReactorHelper.monoError(ex, log));
    }

    /**
     * Tries to get the {@link NetPlayerHistory} data container from the BAGR instance.
     *
     * @param uuid
     * @return
     */
    public Mono<DataResultable<NetPlayerHistory>> loadPlayerHistory(UUID uuid) {
        return netPlayerService.historyFor(WUUID.newBuilder()
                        .setValue(uuid.toString())
                        .build())
                .filter(ReactorHelper.filterResult(log))
                .map(data -> Packeto.unpack(data, NetPlayerHistory.class))
                .doOnNext(result ->
                        result.ifOk(data -> historyCache.put(uuid, data)))
                .onErrorResume(ex -> ReactorHelper.monoError(ex, log));
    }

    public Mono<Resultable> updateData(NetPlayer netPlayer) {
        return netPlayerService.updateData(netPlayer)
                .map(Resultable::convert)
                .onErrorResume(ex -> ReactorHelper.monoError(ex, log));
    }

    public Mono<Resultable> playerOnline(UUID uuid) {
        return netPlayerService.playerOnline(WUUID.newBuilder()
                        .setValue(uuid.toString())
                        .build())
                .map(Resultable::convert)
                .onErrorResume(ex -> ReactorHelper.monoError(ex, log));
    }

    public Mono<Resultable> playerOffline(UUID uuid) {
        return netPlayerService.playerOffline(WUUID.newBuilder()
                        .setValue(uuid.toString())
                        .build())
                .map(Resultable::convert)
                .onErrorResume(ex -> ReactorHelper.monoError(ex, log));
    }

    public DataResultable<Locale> resolveLocale(UUID uuid) {
        final var data = getPlayer(uuid);
        if (data.isFail()) {
            return data.transform();
        }
        return NetUtils.resolveLocale(data.data().getSettings().getLocale());
    }

    public Flux<NetPlayer> playerUpdater() {
        return playerUpdater.asFlux();
    }

    private void subscribeToUpdates() {
        if (updateListener != null) {
            updateListener.dispose();
        }

        updateListener = netPlayerService.subscribeToUpdates(Empty.getDefaultInstance())
                .doOnNext(data -> {
                    final var uuid = UUID.fromString(data.getOwner().getUuid());
                    playerCache.put(uuid, data);
                    playerUpdater.tryEmitNext(data);
                })
                .onErrorResume(ex -> ReactorHelper.fluxError(ex, log))
                .subscribe();
    }

    public record Settings(NetPlayerProvider client) {
        public Mono<Boolean> forKey(UUID uuid, PlayerSettings.Key key) {
            return Mono.fromSupplier(() -> client.playerCache.getIfPresent(uuid))
                    .map(next -> next.getSettings()
                            .getSettingsMap()
                            .get(key.getNumber()))
                    .switchIfEmpty(Mono.just(false));
        }
    }
}

File: client/src/main/kotlin/net/hoz/netapi/client/provider/NetPlayerProvider.kt
/*
 * Copyright 2022 hoz-network.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package net.hoz.netapi.client.provider

import com.github.benmanes.caffeine.cache.Caffeine
import com.google.protobuf.Empty
import com.iamceph.resulter.core.DataResultable
import com.iamceph.resulter.core.Resultable
import com.iamceph.resulter.kotlin.dataResultable
import com.iamceph.resulter.kotlin.resultable
import com.iamceph.resulter.kotlin.unpack
import kotlinx.coroutines.Job
import kotlinx.coroutines.cancel
import kotlinx.coroutines.coroutineScope
import kotlinx.coroutines.flow.MutableSharedFlow
import kotlinx.coroutines.flow.SharedFlow
import kotlinx.coroutines.flow.collect
import kotlinx.coroutines.flow.onEach
import kotlinx.coroutines.launch
import mu.KotlinLogging
import net.hoz.api.data.NetPlayer
import net.hoz.api.data.PlayerHistory
import net.hoz.api.service.NetPlayerServiceGrpcKt
import net.hoz.api.service.SettingsKeyUpdate
import net.hoz.api.service.netPlayerRequest
import net.hoz.netapi.api.Controlled
import net.hoz.netapi.api.onErrorHandle
import net.hoz.netapi.client.util.NetUtils
import net.hoz.netproto.asProto
import org.screamingsandals.lib.player.PlayerMapper.getPlayer
import org.slf4j.Logger
import java.time.Duration
import java.util.*
import javax.inject.Inject

private val log: Logger = KotlinLogging.logger {}

class NetPlayerProvider @Inject constructor(
    private val playerService: NetPlayerServiceGrpcKt.NetPlayerServiceCoroutineStub,
    private val playerUpdater: MutableSharedFlow<NetPlayer>
) : Controlled {
    /**
     * Cache of the player data.
     */
    private val playerCache = Caffeine.newBuilder()
        .expireAfterAccess(Duration.ofMinutes(30))
        .build<UUID, NetPlayer>()

    /**
     * Cache of the player history.
     */
    private val historyCache = Caffeine.newBuilder()
        .expireAfterAccess(Duration.ofMinutes(30))
        .build<UUID, PlayerHistory>()

    private var playerUpdateJob: Job? = null

    override suspend fun initialize() {
        playerUpdateJob?.cancel("Staring new listening for NetPlayer updates")

        playerUpdateJob = coroutineScope {
            launch {
                playerService.subscribeToUpdates(Empty.getDefaultInstance())
                    .onEach { data: NetPlayer ->
                        val uuid = UUID.fromString(data.id)
                        playerCache.put(uuid, data)
                        playerUpdater.tryEmit(data)
                    }
                    .onErrorHandle(log)
                    .collect()
            }
        }
    }

    override fun dispose() {
        playerUpdateJob?.cancel("Ending listening for Player updates.")

        playerCache.invalidateAll()
        historyCache.invalidateAll()
    }

    fun playerUpdater(): SharedFlow<NetPlayer> = playerUpdater

    /**
     * Provides [NetPlayer] for given [UUID]
     */
    fun player(uuid: UUID): NetPlayer? = playerCache.getIfPresent(uuid)

    /**
     * Provides [PlayerHistory] for given [UUID]
     */
    fun history(uuid: UUID): PlayerHistory? = historyCache.getIfPresent(uuid)

    /**
     * Tries to get the [NetPlayer] data container from the BAGR instance.
     */
    suspend fun loadPlayer(uuid: UUID, address: String?): DataResultable<NetPlayer> {
        return playerService.dataFor(
            netPlayerRequest {
                this.id = uuid.toString()
                this.address = address
                    ?: ""
            }
        )
            .unpack(NetPlayer::class)
            .ifOk { playerCache.put(uuid, it) }
    }

    /**
     * Tries to get the [PlayerHistory] data container from the BAGR instance.
     */
    suspend fun loadPlayerHistory(uuid: UUID): DataResultable<PlayerHistory> {
        return playerService.historyFor(uuid.asProto())
            .unpack(PlayerHistory::class)
            .ifOk { historyCache.put(uuid, it) }
    }

    suspend fun updateSettings(settings: SettingsKeyUpdate): Resultable = playerService.updateSettingsByKey(settings)
        .resultable()

    suspend fun playerOnline(uuid: UUID): Resultable = playerService.playerOnline(uuid.asProto())
        .resultable()

    suspend fun playerOffline(uuid: UUID): Resultable = playerService.playerOffline(uuid.asProto())
        .resultable()

    fun resolveLocale(uuid: UUID): DataResultable<Locale> {
        val player = player(uuid)
            ?: return DataResultable.fail("Player not found.")
        return NetUtils.resolveLocale(player.settings.locale)
    }
}